full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,152 @@
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function getAllTodos() {
+      try {
+        const response = await fetch('/api/todos');
+        if (!response.ok) {
+          throw new Error(`HTTP error! Status: ${response.status}`);
+        }
+        const data = (await response.json()) as Todo[];
+        setTodos(data);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    getAllTodos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const response = await fetch('/api/todos/', {
+        method: 'post',
+        body: JSON.stringify(newTodo),
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+      const data = (await response.json()) as Todo;
+      setTodos([...todos, data]);
+    } catch (err) {
+      setError(err);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todo: Todo) {
+    try {
+      const response = await fetch(`/api/todos/${todo.todoId}`, {
+        method: 'put',
+        body: JSON.stringify({
+          task: todo.task,
```

Comment:
Another way to do this, since there may be multiple fields in the object you are modifying, is:
`{ ...todo, isCompleted: !todo.isCompleted }`.